modules/product/src/main/java/com/opengamma/strata/product/SecurityInfo.java

```diff
@@ -0,0 +1,414 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.product;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.Set;
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanBuilder;
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.google.common.collect.ImmutableMap;
+import com.opengamma.strata.basics.market.ReferenceData;
+import com.opengamma.strata.collect.Messages;
+
+/**
+ * Additional information about a security.
+ * <p>
+ * This allows additional information about a security to be provided if desired.
+ * All properties on this class are optional.
+ */
+@BeanDefinition(builderScope = "private")
+public final class SecurityInfo
+    implements ImmutableBean, Serializable {
+
+  /**
+   * The security identifier.
+   * <p>
+   * This identifier uniquely identifies the security within the system.
+   * It is the key used to lookup the security in {@link ReferenceData}.
+   * <p>
+   * A real-world security will typically have multiple identifiers.
+   * The only restriction placed on the identifier is that it is sufficiently
+   * unique for the reference data lookup. As such, it is acceptable to use
+   * an identifier from a well-known global or vendor symbology.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final SecurityId id;
+  /**
+   * The additional security information.
+   * <p>
+   * This stores additional information for the security.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final ImmutableMap<SecurityInfoType<?>, Object> info;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an instance from the identifier.
+   * 
+   * @param id  the security identifier
+   * @return the security information
+   */
+  public static SecurityInfo of(SecurityId id) {
+    return new SecurityInfo(id, ImmutableMap.of());
+  }
+
+  /**
+   * Obtains an instance from the identifier and additional information.
+   * 
+   * @param id  the security identifier
+   * @param infoMap  the map of additional information
+   * @return the security information
+   */
+  public static SecurityInfo of(SecurityId id, Map<SecurityInfoType<?>, Object> infoMap) {
```

Comment:
I'm not sure about this signature. `SecurityInfoType` doesn't have a field containing the `Class` so there is no way check the type at runtime. This map is completely type-unsafe and any problems won't be evident until the data is looked up and a `ClassCastException` is thrown.

If we can't have a `Class` field on `SecurityInfoType` then I think the least we should do is remove this method and have a builder with a method:

    Builder <T> addSecurityInfo(SecurityInfotype<T> type, T info);

That way we have some compile-time type checking of the info values.